feat: Add Google BigQuery data loader (#1543)
observablehq/framework: examples/loader-google-bigquery/src/index.md

  } catch (error) {
    console.error('Error running query or processing file:', error);
  }
})();
```

<div class="tip">

The Google BigQuery API returns dates in ISO 8601 `YYYY-MM-DD` format, which ensures a standard representation that will be correctly parsed by `FileAttachment.csv`.

</div>

The data loader uses a helper file, `google-bigquery.js`, which is a thin wrapper on the `@google-cloud/bigquery` package. This reduces the amount of boilerplate you need to run a report.

```js run=false
import "dotenv/config";
import {BigQuery} from "@google-cloud/bigquery";

const {BQ_PROJECT_ID, BQ_CLIENT_EMAIL, BQ_PRIVATE_KEY} = process.env;

if (!BQ_PROJECT_ID) throw new Error("missing BQ_PROJECT_ID");
if (!BQ_CLIENT_EMAIL) throw new Error("missing BQ_CLIENT_EMAIL");
if (!BQ_PRIVATE_KEY) throw new Error("missing BQ_PRIVATE_KEY");

const bigQueryClient = new BigQuery({
  projectId: BQ_PROJECT_ID,
  credentials: {
    client_email: BQ_CLIENT_EMAIL,
    private_key: BQ_PRIVATE_KEY.replace(/\\n/g, '\n')
  }
});

export async function runQuery(query) {
  const [rows] = await bigQueryClient.query({ query });
  return rows;
}
```

<div class="note">

> To run this data loader, you’ll need to install `@google-cloud/bigquery`, `d3-dsv`, `d3-time-format`, and `dotenv` using your preferred package manager such as npm or Yarn.

Reviewer: ```suggestion
This data loader depends on `@google-cloud/bigquery`, `d3-dsv`, `d3-time-format`, and `dotenv`, which we reference in `package.json`.
```

I wonder if this kind of wording might be better? Since there's nothing special to do for the user besides installing the project.

</div>

<div class="note">

See the [Google Bigquery API Quickstart](https://cloud.google.com/bigquery/docs/authentication) for how to create the service account needed to access the Google BigQuery API after it has been enable in the GCP console. The following Google codelab might be helpful walk-through before getting started: https://codelabs.developers.google.com/codelabs/cloud-bigquery-nodejs#0

</div>

The above data loader lives in `data/covidstats_it.csv.js`, so we can load the data as `data/covidstats_it.csv`. The `FileAttachment.csv` method parses the file and returns a promise to an array of objects, which gets rendered through/into `covid_2020.md`. To create the csv file, please run `node src/data/covidstats_it.csv.js` from the right directory.

```js echo
const covidStats = FileAttachment("covidstats_it.csv").csv({typed: true});
```

The `covidStats` table has two columns: `formatted_date` and `confirmed_cases`. We can display the table using `Inputs.table`.

```js echo
Inputs.table(covidStats)
```

Lastly, we can pass the table to `Plot.plot` to make a simple line chart.

```js echo
display(

>   Plot.plot({

Reviewer: That's a bit too much code for a simple line chart, we can trim it down quite a bit (I can help with that).

Author: Yes, can you trim please 🙏

    width: 800,
    height: 400,
    marginTop: 30,
    marginBottom: 90,  
    x: {
      type: "utc",
      label: "Date",
      labelAnchor: "right", 
      tickFormat: d3.utcFormat("%Y-%m-%d"),
      tickRotate: -90 
    },
    y: {
      grid: true,
      label: "Confirmed Cases",
      domain: [2000, 3000], 
      tickFormat: d => d
    },
    marks: [
      Plot.lineY(covidStats, {
        x: d => new Date(d.formatted_date),
        y: "confirmed_cases",
        stroke: "steelblue",
        curve: "step"
      }),
      Plot.ruleY([0])
    ]
  })
);
```

### The output should look as follows:

> ![image info](covid_stats.png)

Reviewer: I don't think it's necessary to add a static image since the chart is displayed just above?

Author: Agreed :) The reason why I added it, is that it is always night to upfront see what users want to replicate and whether the result what they need to achieve is the same. Not absolutely necessary and might even be better to move this image to the README.
